### ISSUE TITLE
How to configure "endless" re-feed of banned IP to jail action

### ISSUE THREAD

**Issue by joe-average-user** [fail2ban/fail2ban#2911]
As this is not really an "issue" per se but rather a knowledge question of general interest (I suggest :-) let me explain the "problem".

I have a jail set up which consists of three parts with increasing bantime, configured like this:
```ini
[DEFAULT]
bantime.increment = true
bantime = 7200 
bantime.multipliers = 1 12 1080
findtime = 7776000
ignoreip = 127.0.0.1/8 192.168.0.0/16 
maxmatches = 0 

[honeypot]
enabled = true
filter  = honeypot-combined
action  = ban-combined
logpath = /log/honeypot.log
maxretry = 1
```
Action fills 3 ipsets according to the count given at execution with increasing bantime.
As soon as an IP expires from the third ipset and re-appears for ban I see this in the logs:

"Ignore A.B.C.D, expired bantime"

Now, how do I re-ban this already known IP again in the ipset with the highest bantime?
The "action" script is not executed, as the IP is "ignored".
What to do?

**Comment by sebres** [comment]
> Action fills 3 ipsets according to the count given at execution with increasing bantime.
> As soon as an IP expires from the third ipset and re-appears for ban...

Can you please explain this sentences? Can you provide a fail2ban.log excerpt for all messages of such IP?
What exactly re-appears for ban means?
Does fail2ban have found this IP 3rd or 4th time, so the observer (ban-time increment) wanted to increase ban (again) and it expire for some reason?

**Which fail2ban version exactly is it?** 
Does action `ban-combined` use some fail2ban functionality?
Can you provide an excerpt of `fail2ban-client --dp` to know all the parameters (black some params you don't want to publish)?

<hr>

By the way, you seem to have very long failure "buffer" (failure time window due to `findtime = 7776000`, or 90days) - it is really no favours:
- a good "visited" service, which could become a lot of failures from different IPs, they all don't reach `maxretry` would cause extremely high memory consumption (every IP stored as ticket in faimanager with its fail counter etc, in order to consider `maxretry` later, and it gets forgotten not earlier than after 3 months in your case).
- it just unneeded increase a ticket search time in the failmanager (to increase its failure count);
- restart of fail2ban can cause too large scan (causes only an unneeded server load and retards analyze of messages which occurring right now (e. g. causes unneeded delay for intruder attempting brute force now, etc).

**Comment by joe-average-user** [comment]
It's really quite a simple setup Fail2ban version is 0.11.1.
The actions look like this:

actionban   = banip <ip> <bancount> <bantime>
actionprolong = %(actionban)s
actionunban = unban <ip> <bancount> <bantime>

The scripts only use the bancount to set the correct ipset, that's it. There is no fail2ban code involved there.

It seems to me that the number of given bantime.multipliers equals the number of ban actions per IP. I thought I read somewhere that the multipliers for further actions are simply set to the last given multiplier. But instead the logfile only shows an "expired" message and thats it. No more actions are started after the last (in my example third) attempt.
As this logfile shows:
2021-01-14 23:58:20,092 fail2ban.filter         [734]: INFO    [honeypot] Found A.B.C.D - 2021-01-14 23:49:14
2021-01-15 00:00:48,007 fail2ban.filter         [734]: INFO    [honeypot] Found A.B.C.D - 2021-01-14 23:55:22
2021-01-15 04:53:10,977 fail2ban.actions        [734]: INFO    [honeypot] Ignore A.B.C.D, expired bantime

Do I really have to add 50 more multipliers to get more ban tries?

**Comment by joe-average-user** [comment]
FYI: We are not protecting a special service like http or the like. We are harvesting bot net IPs. We found there are about 700.000 bot net IPs currently on the planet. If you simply lock out all these there is merely no spam and no other strange traffic in your net. We do :-)

**Comment by joe-average-user** [comment]
Have to resend this, because the brackets were not transmitted:

actionban   = banip ip bancount bantime
actionprolong = %(actionban)s
actionunban = unban ip bancount bantime

You know where the brackets should be, they are there, but cannot be inserted here.

**Comment by sebres** [comment]
> I thought I read somewhere that the multipliers for further actions are simply set to the last given multiplier.

Yes, it is. But the message "expired bantime" is throw from action and has basically nothing to do with observer module (which prolong the ban asynchronously), so firstly you should anyway see the message `Ban <IP>`. I'll check the case if prolongation takes place before the ban and what happens by such large multiplier...

**Comment by joe-average-user** [comment]
No, there is no ban message. And as far as I read the source this "expired" is a separate case in actions.py line 522ff. There is no further action taken in this case. So explicitely there is no ban action in this case.
Which basically means that such IPs never get banned any more after they "fell out" of the last bantime.

**Comment by joe-average-user** [comment]
There must be another problem going on in this code. I just found out that the "observer" has not made any log entry since 2 days. This is strange because the logfile to observe grows every second. Is there a way to find out what is going on with the observer? Is there some error condition I can find out about?

**Comment by sebres** [comment]
No, "observer" is an asynchronous module observing bans and findings of jails/filters (to increase ban-time and decrease maxretry for known as bad "attackers"), so if no ban occurs (because of some error here with "expired"), no ban prolongation takes place (observer doesn't get an event).

Back to the issue - I'm pretty sure failures with too long distance due to `findtime`, which is extremely larger as a `bantime` causing that somehow, and that it is fixed already in 0.11.2 (where I was unable to reproduce that)... but to provide an evidence I need to write some test which should fail in 0.11.1 (generate an expired) and succeed in 0.11.2 (and I still don't have such test ready). 
And to tell you test it were wrong, because with your config you'd need to wait up to 3 months to get the same conditions.

**Comment by sebres** [comment]
By the way, is it released 0.11.1, or some development state (0.11.1-dev)?

**Comment by joe-average-user** [comment]
what would you say : 538660 Mar 13  2020 fail2ban-0.11.1.tar.gz ?

**Comment by joe-average-user** [comment]
Are you saying it would be the best and most simple solution to update to 0.11.2 ?

**Comment by sebres** [comment]
and where is the crash?

**Comment by joe-average-user** [comment]
The log says exactly nothing about the crash. Only trying to talk to the server reveals that it does not seem to answer. On startup it does about 700 bans, and then its dead.

**Comment by sebres** [comment]
don't even journal saying something about crash or what is happening? `systemctl status fail2ban` or `journalctl -r`

**Comment by joe-average-user** [comment]
This is no systemd setup. Its a pretty old init style thing. I have not expected that the difference between 0.11.1 and 0.11.2 is that big ...

**Comment by joe-average-user** [comment]
How do I re-install 0.11.1 over 0.11.2 because "python setup.py install"  does not seem to work ...

**Comment by sebres** [comment]
> the difference between 0.11.1 and 0.11.2 is that big

It is really not. 
I don't know what happens there... Excepting that by findtime 3 month (7776000) and no log rotation, I could imagine that fail2ban would find a lot of messages which all can overflow your memory (even with `maxmatches = 0`), just too many different IPs. 
If so switch to 0.11.1 would probably change nothing. It was rather a restart then.

**Comment by joe-average-user** [comment]
No, the restart is no problem. I have restarted fail2ban about 15 times during the last 3 days. It never had problems with that.

**Comment by sebres** [comment]
> How do I re-install 0.11.1?

How you installed it?

>  "python setup.py install" does not seem to work ...

What is the message?
If your fail2ban version was made for 3.x, you have to use `python3 setup.py install` instead.

**Comment by sebres** [comment]
Before reinstall, can you probably try to start from console in foreground mode (in hope it would again crash and one sees some message there)?
```
?sudo? fail2ban-server -f
```

**Comment by joe-average-user** [comment]
No, this is a python2.6.
I tried just like former install was: "python setup.py install". Then it shows:
# python setup.py install
running install
running build
running build_py
running build_scripts
running install_lib
running install_scripts
WARNING: Cannot find root-base option, check the bin-path to fail2ban-scripts in "fail2ban.service".
Creating build/fail2ban.service (from fail2ban.service.in): @BINDIR@ -> /usr/local/bin
creating fail2ban-python binding -> /usr/local/bin
changing mode of /usr/local/bin/fail2ban-client to 755
changing mode of /usr/local/bin/fail2ban-testcases to 755
changing mode of /usr/local/bin/fail2ban-server to 755
changing mode of /usr/local/bin/fail2ban-regex to 755
running install_data
running install_egg_info
Removing /usr/local/lib/python2.6/site-packages/fail2ban-0.11.1-py2.6.egg-info
Writing /usr/local/lib/python2.6/site-packages/fail2ban-0.11.1-py2.6.egg-info

Please do not forget to update your configuration files.
They are in "/etc/fail2ban/".

You can also install systemd service-unit file from "build/fail2ban.service"
resp. corresponding init script from "files/*-initd".


But if I try "fail2ban-client start" afterwards, fail2ban crashes just like 0.11.2

**Comment by sebres** [comment]
So it is not 0.11.2, it is something different.

You can check the version with `fail2ban-client --version`.

**Comment by joe-average-user** [comment]
Ok, I deleted the fail2ban in python and re-installed, and now it works. Seems the install does not overwrite already existing files (which you see above). Now it re-created the python fail2ban dir and works again. No crash from server...

**Comment by joe-average-user** [comment]
Is there some major difference between 0.11.1 and 0.11.2 regarding the used python functions? my python 2.6 is not exactly young ..

**Comment by joe-average-user** [comment]
Honestly Sergey, I am very thankful for your work and help during the last years. But somehow I have the impression that what I am trying to do here is so simple that fail2ban is probably a lot oversized for the job. Remember, all I want is some daemon that reads a text file, draws the SRC=<IP> from it, looks into a database how often it has seen this IP and then calls a ban script with this IP and COUNT as arguments. There is no scanning of the text file necessary as that is done elsewhere already. Just take IP, count it, ban it - and release it some time later. I thought this should be very simple, only big in number of IPs, but not more.
Do you think fail2ban is the right thing to use here?
I fell into a problem already some days ago where I found this expression not to work:

failregex = SRC=<HOST>

instead I had to use:

failregex = ^.* core-.* kernel: .* SRC=<HOST> .*$

because fail2ban did accept lines with "core-1 kernel:" in it, but not "core-2 kernel:".
I failed to understand why.

**Comment by sebres** [comment]
> Is there some major difference between 0.11.1 and 0.11.2 regarding the used python functions?

No. But it is hard to say by such usage like yours (and 2.6). 

fail2ban is python pure and all the binaries are the python itself, pyinotify, python-systemd and some standard modules.
I don't know what can cause it, and cannot help here without any info (it crashes is not helpful at all).
If you can provide more data (e.g. could start it in foreground as I requested, with some dummy jail), you can do it parallel to your running instance without installation.

copy your config and database into some temp folder (replace the action with `dummy` to avoid double banning) and follow https://github.com/fail2ban/fail2ban/wiki/How-to-test-newer-fail2ban-version-resp.-use-fail2ban-standalone-instance
So you can start and test it without to bother your other instance, without to real ban-action, etc.

Probably then we'll see some crash report in stderr or in log.

**Comment by joe-average-user** [comment]
I can do the foreground run with the real production setup, as I can delay its function. How do I do that (quick and dirty) ?

**Comment by sebres** [comment]
> I fell into a problem already some days ago where I found this expression not to work...

Both REs are totally non-optimal (if one should say weird).
Since host allows to match hostname it can really match something other as SRC=IP... but dunno without example.
But I would replace `<HOST>` with `<ADDR>` anyway (it is more precise).

> because fail2ban did accept lines with "core-1 kernel:" in it, but not "core-2 kernel:".

Can you give me an example of both messages (with core1 and core2, so working and not).

> with the real production setup ... quick and dirty

```
mkdir /tmp/tests & cd /tmp/tests
git clone --branch 0.11 https://github.com/fail2ban/fail2ban.git /tmp/tests/f2b
cd /tmp/tests/f2b
PYTHONPATH=. ./bin/fail2ban-server -f
```
I don't know faster way :)

it will use `/etc/fail2ban` as config, if you will run it parallelly you can copy it to local `config`, adjust it and add `-c /tmp/tests/f2b/config` to command line to use different data & conf for that instance.

**Comment by joe-average-user** [comment]
Ok, it seems the foreground run does not crash. At least not that fast. Can this be a stack overflow or something the like?

**Comment by sebres** [comment]
It can be everything :) But sure, one could compare the limits of process in console and that as daemon.

**Comment by joe-average-user** [comment]
Ok, an example of some log line:

Jan 21 17:25:09 core-2 kernel: IN=eth3 OUT=eth0 MAC=52:54:00:8d:eb:41:24:6e:96:09:6e:c4:08:00 SRC=A.B.C.D DST=E.F.G.H LEN=40 TOS=0x00 PREC=0x00 TTL=245 ID=2811 DF PROTO=TCP SPT=12345 DPT=12 WINDOW=14600 RES=0x00 SYN URGP=0 

The very same line with "core-1" instead of "core-2" was processed, the above not.

**Comment by joe-average-user** [comment]
> It can be everything :) But sure, one could compare the limits of process in console and that as daemon.

How can that be different using 0.11.1 and 0.11.2 ? Both installed from scratch can only mean that 0.11.2 simply needs more space somehow. How can I enlarge the limits for the spawned server process?
(Sorry if that sounds dumb, but I thought they should be inherited...?)

**Comment by sebres** [comment]
```
fail2ban-regex -v 'Jan 21 17:25:09 core-2 kernel: IN=eth3 OUT=eth0 MAC=52:54:00:8d:eb:41:24:6e:96:09:6e:c4:08:00 SRC=192.0.2.111 DST=192.0.2.1 LEN=40 TOS=0x00 PREC=0x00 TTL=245 ID=2811 DF PROTO=TCP SPT=12345 DPT=12 WINDOW=14600 RES=0x00 SYN URGP=0' 'SRC=<HOST>'
```
matches it without any issue, no matter `core-2` or `core-1`. Something else (e. g. prefregex, ignoreregex or so) is in filter or jail that can avoid that. Check the config dump `fail2ban-client --dp` whether one have something else.

> can only mean that 0.11.2 simply needs more space somehow.

Well in case of stack overflow it could call some function (enter one level more in stack) to overflow that. 
But I cannot really remember any "big" change which could cause a crash.

> How can I enlarge the limits for the spawned server process?

1. I thought you started it as init.d daemon, don't you? In this case it inherits surely from init process (and it may have different limits).
2. you can specify the limits per user/group also (so root has another than user, you tried to start fail2ban from a console).

**Comment by joe-average-user** [comment]
Ok, back to work :-)
What I don't understand is this: starting fail2ban 0.11.2 with "fail2ban-client start" from console lets f2b-server process crash. Whereas starting "fail2ban-server -f" from the very same console works perfectly.
Why this?

**Comment by joe-average-user** [comment]
Update: Running 0.11.2 in foreground for around 12 hours I must say it works very well. I changed the filter statement as suggested and it works. There is no delay between the fed logfile and the f2b reaction. All IPs I checked so far are really processed.
The only thing that's left is the problem with starting it in the background by "fail2ban-client start". What limits are inherited in this case? Something must be different from the starting in the foreground...

**Comment by sebres** [comment]
> Staring with "fail2ban-client start" from console lets f2b-server process crash. Whereas starting "fail2ban-server -f" from the very same console works perfectly.

`fail2ban-client start` would create a fork, configure it (read a config in clients, generate a stream and supply to the server) and hereafter returns to the shell, 
whereas `fail2ban-server -f` starts server immediately (and an asynchronous worker reads config and do all the stuff that client do in above)... and it is still running further (so single instance self-configuring process).
So the sole difference here is the fork (and probably the kind of configure process, how the config-stream arrives the server).
But there are no differences **in this logic** between 0.11.1 and 0.11.2.

The only one notable difference here is [this](https://github.com/fail2ban/fail2ban/compare/0.11.1...0.11#diff-b7dd4978b366da82d24bab21cfe80a7718329bff4f3f23fa4dcb0f8e534a2f02R132-R133) belonging to 3f489070646b363aa0374681fe910f05521cd247 (now it is moved from constructor to start and main worker renamed from "f2b/server" to "fail2ban-server"). But although I cannot say how it can affect that (in such old python version), I don't think it causing this bug (because you said it is running short time, banned 700 tickets, but if it were the bug it should cause the crash almost immediately after start of fork).

Another thing that could affect command-line processing may be 5abc4ba4ae280b6ec89c6bffc9dd2140d0d56dc4 (fix for #2758) which changing the reaction and the behavior how the standard pipes gets closed now by exit, but I can also not follow the relation to the crash here (the handles of fork should be duplicated or not available at all, so client can properly close that after fork started).

I don't really know how exactly forking may affect the remaining changes on modules in 0.11.2, but if I do a review of modification I'm really seeing nothing what could cause such a crash.
As for memory an unforked variant (without fail2ban-client), it should even consume more memory at start (because client and server functionality in the same process, whereas with client they are separated in two processes), and after the start (configuration is ready) it does no matter anymore.

Unfortunately, I cannot help you here as I really have nothing (no callstack/backtrace, no info about segfault, or out of limits, etc) about what it exactly is or may be (but it smells quasi like it would be python2.6 issue).
So the solutions would be to try:
- newer python version (you can still remain by 2.x, but why 2.6 and not 2.7, which received soooo many fixes against 2.6?)
- another variant to start it as daemon without fork, e. g. change init.d, so it would start the server via `fail2ban-server -f &` (note `&` must be last parameter after all other arguments your fail2ban server instance needed);
- to investigate deeper with dgb of forked process and/or redirect its standard pipes to some file to see what is happening there.

**Comment by sebres** [comment]
I guess we can close it here, don't hesitate to ping me if you mean the issue must be reopened (but I don't really know what we can help here further by this second issue with crashing of 0.11.2 in python 2.6, especially if some configuration constellation or module is involved).
